syntax = "proto3";

package com.empuje.inventoryservice.grpc.gen;

import "google/protobuf/timestamp.proto";

option java_multiple_files = true;
option java_package = "com.empuje.inventoryservice.grpc.gen";
option java_outer_classname = "InventoryServiceProto";

service InventoryService {
  rpc AddItem(AddItemRequest) returns (ItemResponse) {}
  rpc UpdateItem(UpdateItemRequest) returns (ItemResponse) {}
  rpc DeleteItem(DeleteItemRequest) returns (DeleteItemResponse) {}
  rpc ListItems(ListItemsRequest) returns (stream ItemResponse) {}
  rpc AdjustQuantity(AdjustQtyRequest) returns (ItemResponse) {}
}

enum Category {
  CATEGORY_UNKNOWN = 0;
  ROPA = 1;
  ALIMENTOS = 2;
  JUGUETES = 3;
  UTILES_ESCOLARES = 4;
}

message AddItemRequest {
  Category category = 1;
  string description = 2;
  int32 quantity = 3;
  int64 created_by = 4;
}

message UpdateItemRequest {
  int64 id = 1;
  string description = 2;
  int32 quantity = 3;
  int64 updated_by = 4;
}

message DeleteItemRequest {
  int64 id = 1;
  int64 updated_by = 2;
}

message DeleteItemResponse {
  bool success = 1;
  string message = 2;
}

message AdjustQtyRequest {
  int64 id = 1;
  int32 delta = 2;
  int64 updated_by = 3;
}

message ItemResponse {
  int64 id = 1;
  Category category = 2;
  string description = 3;
  int32 quantity = 4;
  bool deleted = 5;
  google.protobuf.Timestamp created_at = 6;
}

message ListItemsRequest {}
